extensions/amp-fit-text/0.2/amp-fit-text.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {FitText} from './fit-text';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {getLengthNumeral, isLayoutSizeDefined} from '../../../src/layout';
+import {isExperimentOn} from '../../../src/experiments';
+
+import {dict} from '../../../src/utils/object';
+import {userAssert} from '../../../src/log';
+
+/** @const {string} */
+const TAG = 'amp-fit-text';
+
+const getFontSizeAttrs = (element) =>
+  dict({
+    'maxFontSize': getLengthNumeral(element.getAttribute('max-font-size')),
+    'minFontSize': getLengthNumeral(element.getAttribute('min-font-size')),
+  });
+
+class AmpFitText extends PreactBaseElement {
+  /** @override */
+  init() {
+    const attributeOb = new MutationObserver(() => {
+      this.mutateProps(getFontSizeAttrs(this.element));
+    });
+    attributeOb.observe(this.element, {
+      attributeFilter: ['min-font-size', 'max-font-size'],
+    });
+
+    // Force render to resize to new contents.
+    const childOb = new MutationObserver(() => {
+      this.mutateProps(dict({}));
+    });
+    childOb.observe(this.element, {childList: true});
```

Comment:
This should be `subtree` observer I believe to deal with `<ul>` and other markup nuances. And we should also limit it to the attributes we look for in children. E.g. only with "option" and "selected", right?

Reply:
Oops. I shouldn't be reviewing both amp-fit-text and amp-selector at the same time. `attributesFilter` is not needed here. However, this should still be `subtree`, right?

Reply:
Ya, good catch. But I think we need to resize on any childlist mutation? That could include child attr mutations, too.

Reply:
To summarize:
- Add `childList` for changes in the target node
- Add `subtree` for changes on descendants of the target node
- Add `attributes: true` for attr changes all along this tree ^

Reply:
Is there any need to throttle the callback in case there are many DOM events occurring in quick succession? Does the browser handle this?

Reply:
I'd not throttle, browser manages the load for us: we will just have several mutation records in the array - though we don't really care how many of them are - one or many we do the same thing.

Reply:
Having both `subtree` and `attributes` present is causing the observer callback to fire infinitely due to the `i-amphtml-context` attribute being re-applied on every render. Going to remove one of these for now.

Reply:
Can we simply ignore mutation records with `attribute = i-amphtml-context`? Or even with any `startsWith(attribute, 'i-amphtml-')`?

Reply:
Yes, would it be costly to filter every set of mutations for this?